### PR TITLE
Package show URI

### PR DESCRIPTION
**Pull request by thedaniel**
This allows a package detail view to be shown directly from a URI, whether it's installed or not. This is useful for e.g. installing packages from links on the webiste (atom/atom#5262) or linking directly to the the metrics package from atom/welcome to make it simple to disable (atom/atom#4966) among other things. 

Here's a dump of my notes for the flow, it should match the specs as well.

```
# PackageDetailView:

- Is pack present?
  - Yes
    - is it complete?
      - Yes
        - Does everything render normally? Cool.
      - No
        - [A] Is the network available?
          - Yes
            - Call the network.
              - Failure?
                - Show an error page
              - Success?
                - Render that thing (don't forget to rerender the subview)
  - No
    - See [A]
```

### CODE REVIEW

**Comment by thedaniel** [comment]
Anywho this PR represents the design decision that `PackageDetailView` is responsible for figuring out whether the whole package is there and hitting the network and displaying the right state. An alternative is handling it in `main.coffee` or `SettingsView` but this feels right to me.

**Comment by thomasjo** [comment]
Can't wait! Thanks for working on this @thedaniel.

**Comment by thedaniel** [comment]
@atom/feedback i think this is ready to review. I'm not fully confident that it handles all error states but I've tried to handle network errors in a simple, catch-all kind of way.

I'd like to leave this undocumented for now as the initial use is just intended to be https://github.com/atom/welcome/pull/38 - you can link that branch in dev mode to see it in action from the welcome package.

When we address https://github.com/atom/atom/issues/5262 we may discover some necessary changes, etc, and I would suggest documenting this at that time.

File: lib/package-card.coffee [hunk 1]
```diff
@@ -174,9 +174,11 @@ class PackageCard extends View
       @avatar.attr 'src', "file://#{avatarPath}" if avatarPath
 
     @client.package @pack.name, (err, data) =>
-      data ?= {}
-      @packageData = data
-      @downloadCount.text data.downloads?.toLocaleString()
+      if err
```
Comment (benogle) [inline lib/package-card.coffee]: prolly should use an `unless` here

File: spec/package-detail-view-spec.coffee [hunk 2]
```diff
@@ -0,0 +1,71 @@
+fs = require 'fs'
```
Comment (benogle) [inline spec/package-detail-view-spec.coffee]: yay for adding specs to these views

File: spec/package-detail-view-spec.coffee [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    packageManager.client = createClientSpy()
+    packageManager.client.package.andCallFake (name, cb) ->
+      packageData = require(path.join(__dirname, 'fixtures', 'package-with-readme', 'package.json'))
+      packageData.readme = fs.readFileSync(path.join(__dirname, 'fixtures', 'package-with-readme', 'README.md'), 'utf8')
+      cb(null, packageData)
+    view = new PackageDetailView({name: 'package-with-readme'}, packageManager)
+    view.beforeShow(opts)
+
+
+  it "Renders a package when provided in `initialize`", ->
```
Comment (benogle) [inline spec/package-detail-view-spec.coffee]: I think we lowercase the beginning of these.

File: spec/package-detail-view-spec.coffee [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    opts ?= {}
+    packageManager.client = createClientSpy()
+    packageManager.client.package.andCallFake (name, cb) ->
+      packageData = require(path.join(__dirname, 'fixtures', 'package-with-readme', 'package.json'))
+      packageData.readme = fs.readFileSync(path.join(__dirname, 'fixtures', 'package-with-readme', 'README.md'), 'utf8')
+      cb(null, packageData)
+    view = new PackageDetailView({name: 'package-with-readme'}, packageManager)
+    view.beforeShow(opts)
+
+
```
Comment (benogle) [inline spec/package-detail-view-spec.coffee]: :art: maybe one newline here?

**Comment by benogle** [comment]
Will be cool to have this :+1:

**Comment by thedaniel** [comment]
I'm going to merge this after our next release goes out, likely tuesday, and then get that metrics update in right after.

**Comment by thomasjo** [comment]
Been testing this for a couple of days now, and not noticed any regressions or bugs :sparkles: 

Only thing missing, but that could perhaps be addressed in a separate PR, is serializing the current package panel details. If I e.g. reload Atom while viewing a package's config, I'll land back in the core settings panel after the reload. Whereas I expect to still be viewing the package config panel. At the very least I expect to remain in the *Packages* section.

**Comment by thedaniel** [comment]
@thomasjo given that settings-view doesn't serialize that on master either (in other words, it's not a regression here), a separate issue / PR is fine.

**Pull request merged**